### DISCUSSION TITLE
@model_serializer and union types

### DISCUSSION THREAD

**Discussion by mfulgo** [pydantic/pydantic#8541] [Pydantic V2]
This popped up in our switch from pydantic V1 to V2 (`v2.5.3`). We were previously overriding the `.dict()` method to customize the serialization, but our attempt to switch to `@model_serializer` started giving errors like the one produced by this example.

```python
from pydantic import BaseModel, model_serializer


class A(BaseModel):
    a: str


class B(BaseModel):
    b: str

    @model_serializer
    def my_serializer(self):
        return self.b


class Container(BaseModel):
    item: B | A  # A | B seems to work fine


if __name__ == '__main__':
    print(Container(item=B(b="yay")).model_dump_json())
    print(Container(item=A(a="hi")).model_dump_json())
```

Running this, I get (line wrapping added):
```
{"item":"yay"}
Traceback (most recent call last):
  File ".../example.py", line 22, in <module>
    print(Container(item=A(a="hi")).model_dump_json())
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../python3.11/site-packages/pydantic/main.py", line 352, in model_dump_json
    return self.__pydantic_serializer__.to_json(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pydantic_core._pydantic_core.PydanticSerializationError:
 Error serializing to JSON: PydanticSerializationError:
  Error calling function `my_serializer`:
   AttributeError: 'A' object has no attribute 'b'
```

What's confusing to me is that `my_serializer` gets called, but even more so, that it's called with `self` as an instance of `A`, which seems to violate assumptions/conventions about types.

Does the `@model_serializer` need to be written differently (using the `wrap` mode)? Is there a more nefarious bug here?

**Comment by augustebaum** [comment]
Getting a similar issue with Pydantic 2.8.2
